### ISSUE TITLE
Improving error messages

### ISSUE THREAD

**Issue by MaxGabriel** [yesodweb/persistent#849]
A common problem my coworker learning Haskell/Yesod/Persistent is that he's missing a `runDB` in a Yesod Handler function, which gives this kind of error:

```
    • Couldn't match type 'ReaderT backend0 m0' with 'HandlerFor App'
      Expected type: HandlerFor App (Entity Organization)
        Actual type: ReaderT backend0 m0 (Entity Organization)
```

His main complaint, which I think is reasonable, is that nothing says "this is related to Persistent" or "this is related to the database". Could Persistent do anything to give better error messages in this case? Maybe renaming usages of `backend` to `databaseBackend`? Perhaps a type synonym for `ReaderT backend m` used in Persistent would help?

```haskell
insertUniqueEntity
    :: (MonadIO m
       ,PersistRecordBackend record backend
       ,PersistUniqueWrite backend)
    => record -> ReaderT backend m (Maybe (Entity record))
```

I haven't used them before, but it looked like GHC's `TypeError` wouldn't work for this? I tried this which I didn't really think would work (and didn't, but also gave lots of warnings), but I'm really fumbling in the dark with that feature—couldn't find very good documentation on it. 

```
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}

import qualified GHC.TypeLits as GHC 

instance GHC.TypeError (GHC.Text "Yesod handlers cannot call database functions directly, use runDB.")
        => HasPersistBackend (HandlerFor site a) where
  persistBackend = error "unreachable"

instance GHC.TypeError (GHC.Text "Yesod handlers cannot call database functions directly, use runDB.")
        => IsPersistBackend (HandlerFor site a) where
```

**Comment by parsonsmatt** [comment]
`SqlPersistT` is "sort of" the type you ask for,  but the problem is that GHC will report the underlying type. We'd need a `newtype` to provide reliably better error messages.

**Comment by MaxGabriel** [comment]
Yeah, and wrapping everything in a `newtype` would be a breaking change for most all users I think

**Comment by kevin-meyers** [comment]
As a learner I ran into this problem and couldn't figure it out! But your post here helped me fix it. Definitely could've benefited from an improved error

**Comment by danbroooks** [comment]
https://github.com/yesodweb/persistent/issues/1038 is related
